### PR TITLE
Improving performance

### PR DESCRIPTION
**Pull request by mortonjt**
Replaced some of the for loops with _.each statements

Not sure exactly how to test for changes to g_comparisonPositions ...

### CODE REVIEW

File: emperor/support_files/js/emperor.js [hunk 1]
```diff
@@ -728,9 +728,10 @@ function colorChangedForTaxaSpheres(color){
 function colorChangedForTaxaLabels(color){
         // get the taxonomic assignments and append '_taxalabel' to
         // retrieve all the labels belonging to a sphere in the plot
-        for(var key in g_taxaPositions) {
+        _.each(_.range(g_taxaPositions.length),function(key){
+        //for(var key in g_taxaPositions) {
        	        $('#'+key+"_taxalabel").css('color', color);
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: Can you put spaces between + and string/vars? I think it makes it more readable .

File: emperor/support_files/js/emperor.js [hunk 2]
```diff
@@ -873,7 +874,7 @@ function displayTaxaLabels(taxaLevel){
 
         $('#taxalabels').css('display','block');
         $("#taxaLevel" ).html(taxonomy[taxaLevel]);
-        for(var key in g_taxaPositions){
+        _.each(_.range(g_taxaPositions.length),function(key){
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: Another way to do this is by:
_.each(g_taxaPositions, function(your_object, key){... })

_each will return 1: the object, 2: the enumerate, 3: the full object.

File: emperor/support_files/js/emperor.js [hunk 3]
```diff
@@ -1441,7 +1443,8 @@ function drawTaxa(){
 	var whiteColor = new THREE.Color();
 	whiteColor.setHex("0xFFFFFF");
 
-	for (var key in g_taxaPositions){
+	//for (var key in g_taxaPositions){
+        _.each(_.range(g_taxaPositions.length),function(key){
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: Same as above.

File: emperor/support_files/js/emperor.js [hunk 4]
```diff
@@ -1815,7 +1818,19 @@ function changeAxesDisplayed() {
   // 		}
 	// }
 
-	var checkedboxes = [];
+    comparisonPositionlength = Object.keys(g_comparisonPositions).length
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: This is not gonna work cause g_spherePositions is not an object of objects any more. I think we need to change the loop to loop on g_spherePositions and then check if g_comparisonPositions has the same [sampleKey][j].

File: emperor/support_files/js/emperor.js [hunk 5]
```diff
@@ -2393,7 +2408,8 @@ $(document).ready(function() {
 		// this is something that will only happen when drawing biplots
 		if(document.biplotoptions){
 			if(document.biplotoptions.elements[0].checked){
-				for(var key in g_taxaPositions) {
+				//for(var key in g_taxaPositions) {
+				_.each(_.range(g_taxaPositions.length),function(key){
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: Same as above.

File: emperor/support_files/js/emperor.js (hunk 4 above)
Comment (mortonjt) [inline emperor/support_files/js/emperor.js]: I think that makes sense.

If this isn't going to work - is there some test case to show that this is broken?

Comment (antgonza) [inline emperor/support_files/js/emperor.js]: No, AFAIK, which makes it really difficult.

Comment (mortonjt) [inline emperor/support_files/js/emperor.js]: OK.  Let's add one in then

**Comment by ElDeveloper** [comment]
This looks good to me!

**Comment by mortonjt** [comment]
@ElDeveloper, https://github.com/antgonza/emperor/pull/1#discussion-diff-26717574R1821 is still a problem.  I'm not sure how to test it.  Could you take a look at it?

**Comment by ElDeveloper** [comment]
I'm having a look right now!

**Comment by mortonjt** [comment]
Some of the tests are broken ... let me fix them first

**Comment by mortonjt** [comment]
Scratch that last statement  -  I got mixed up with git.  The unittests passed.

So that takes care of most of the for loops except for handling
- comparison plots
- animations

**Comment by ElDeveloper** [comment]
Sweet, I'll work on the comparison plots and animations. Do we want to merge this for now then?

File: emperor/support_files/js/emperor.js [hunk 6]
```diff
@@ -2257,15 +2271,15 @@ $(document).ready(function() {
 
 		// build divs to hold point labels and position them
 		var labelshtml = "";
-		for(var i in g_plotIds) {
-			var sid = g_plotIds[i];
+	        _.each(g_plotIds, function(pltID){
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: Why not use _.each(g_plotIds, function(sid){ and remove the next line. This also applies to the other instances.

**Comment by mortonjt** [comment]
I have permissions to do the merge.  @antgonza does that sound good?

**Comment by antgonza** [comment]
Just 1 more comment.

**Comment by antgonza** [comment]
Nop!

**Comment by mortonjt** [comment]
Good thing we are doing PRs right?

File: emperor/support_files/js/emperor.js (hunk 6 above)
Comment (mortonjt) [inline emperor/support_files/js/emperor.js]: Fixed

File: emperor/support_files/js/emperor.js [hunk 7]
```diff
@@ -873,15 +873,15 @@ function displayTaxaLabels(taxaLevel){
...
         $("#taxaLevel" ).html(taxonomy[taxaLevel]);
-        for(var key in g_taxaPositions){
+        _.each(g_taxaPositions, function(taxaPos, key){
             // get the coordinate of this taxa sphere
             var coords = toScreenXY(g_plotTaxa[key].position,g_sceneCamera,$('#main-plot'));
 
-            // labels are identified by the key they have in g_taxaPositions
-            labelshtml += "<label id=\""+key+"_taxalabel\" class=\"unselectable labels\" style=\"position:absolute; left:"+parseInt(coords['x'])+"px; top:"+parseInt(coords['y'])+"px;\">";
-            labelshtml += truncateLevel(g_taxaPositions[key]['lineage'], taxaLevel);
+            // labels are identified by the key they have in taxaPos
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js:881]: I think changing this for append (as in the other _.each) will be nicer. Note that you will need to remove the document... line after the each.

File: emperor/support_files/js/emperor.js [hunk 8]
```diff
@@ -1441,19 +1442,19 @@ function drawTaxa(){
 	var whiteColor = new THREE.Color();
 	whiteColor.setHex("0xFFFFFF");
 
-	for (var key in g_taxaPositions){
+        _.each(g_taxaPositions, function(taxaPos, key){
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js:1448]: You do not need key anymore.

**Comment by ElDeveloper** [comment]
Yeah, peer-review is awesome!

File: emperor/support_files/js/emperor.js [hunk 9]
```diff
@@ -2157,31 +2170,32 @@ $(document).ready(function() {
...
 			var temp = [];
-			for(var j in g_plotIds) {
-				if(g_mappingFileData[g_plotIds[j]] == undefined){
-					console.warning(g_plotIds[j] +" not in mapping")
-					continue
-				}
-				temp.push(g_mappingFileData[g_plotIds[j]][i])
-			}
+                        _.each(g_plotIds, function(pltIds){
+			    if(g_mappingFileData[pltIds] == undefined){
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: Do we really need this console warning?

File: emperor/support_files/js/emperor.js [hunk 10]
```diff
@@ -2157,31 +2170,32 @@ $(document).ready(function() {
...
+				temp.push(g_mappingFileData[pltIds][i]);
+			    }
+			});
 			temp = _.uniq(temp, false);
 
 			// note that each category is added to all the dropdown menus in the
 			// user interface, these are declared in _EMPEROR_FOOTER_HTML_STRING
 			if (i==0) {
-			    line = "<option selected value=\""+sortedMappingFileHeaders[i]+"\">"+sortedMappingFileHeaders[i]+"</option>"
+			    line = "<option selected value=\""+sortedMapHeaders+"\">"+sortedMapHeaders+"</option>"
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: spaces between + and ...

File: emperor/support_files/js/emperor.js [hunk 11]
```diff
@@ -2157,31 +2170,32 @@ $(document).ready(function() {
...
 			} else {
-			    line = "<option value=\""+sortedMappingFileHeaders[i]+"\">"+sortedMappingFileHeaders[i]+"</option>"
+			    line = "<option value=\""+sortedMapHeaders+"\">"+sortedMapHeaders+"</option>"
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: same here

File: emperor/support_files/js/emperor.js (hunk 9 above)
Comment (mortonjt) [inline emperor/support_files/js/emperor.js]: I wasn't sure - it was in there before

File: emperor/support_files/js/emperor.js [hunk 12]
```diff
@@ -2206,10 +2220,10 @@ $(document).ready(function() {
 			$("#gradient-category-drop-down").append(line);
 		}
 
-        for (var i in k_CHROMABREWER_MAPS){
-			line = '<option value="'+k_CHROMABREWER_MAPS[i]+'">'+k_CHROMABREWER_MAPNAMES[i]+'</option>';
+	_.each(k_CHROMABREWER_MAPS, function(map){
+	    line = '<option value="'+ map +'">'+ map +'</option>';
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: spaces?

File: emperor/support_files/js/emperor.js [hunk 13]
```diff
@@ -2257,15 +2271,14 @@ $(document).ready(function() {
 
 		// build divs to hold point labels and position them
 		var labelshtml = "";
-		for(var i in g_plotIds) {
-			var sid = g_plotIds[i];
+	        _.each(g_plotIds, function(sid){
```
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: Same about append.

File: emperor/support_files/js/emperor.js (hunk 8 above)
Comment (mortonjt) [inline emperor/support_files/js/emperor.js:1448]: you sure about that? `g_plotTaxaArrows` is still being indexed by `key`

Comment (antgonza) [inline emperor/support_files/js/emperor.js:1448]: You are right, missed that last line.

File: emperor/support_files/js/emperor.js (hunk 9 above)
Comment (antgonza) [inline emperor/support_files/js/emperor.js]: I would say erase it.

**Pull request merged**